### PR TITLE
Optimize MvcRouteHandler

### PR DESCRIPTION
**Pull request by rynowak**
- Check MVC services once at startup
- Make action selector sync

We've never really had a scenario for the action selector being async, it
just ended up that way. None of our extensibility here lets you do
anything async without replacing it wholesale, which we don't
recommend.

### CODE REVIEW

**Comment by rynowak** [comment]
Here's the data that I'm working with:

image

We can get rid of a state machine here by making MvcRouteHandler just be synchronous (and `Task`-returning.

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/IActionSelector.cs [hunk 1]
```diff
@@ -9,6 +8,6 @@ namespace Microsoft.AspNet.Mvc.Infrastructure
 {
     public interface IActionSelector
     {
-        Task<ActionDescriptor> SelectAsync(RouteContext context);
+        ActionDescriptor SelectAsync(RouteContext context);
```
Comment (Tragetaschen) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/IActionSelector.cs]: Shouldn't this lose the `Async` suffix now as well?

Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/IActionSelector.cs]: :+1:

Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/IActionSelector.cs]: And same for all the other methods that are now sync.

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/MvcRouteHandler.cs [hunk 2]
```diff
@@ -68,12 +63,13 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
                         context.RouteData.Values.Add(kvp.Key, kvp.Value);
                     }
                 }
-            }
 
-            // Removing RouteGroup from RouteValues to simulate the result of conventional routing
-            context.RouteData.Values.Remove(TreeRouter.RouteGroupKey);
+                // Removing RouteGroup from RouteValues to simulate the result of conventional routing
+                context.RouteData.Values.Remove(TreeRouter.RouteGroupKey);
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/MvcRouteHandler.cs]: Any particular reason this moved to inside the `if()`? Is it not needed in other cases?

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/MvcRouteHandler.cs]: Yes, it's an attribute-routing-only kinda thing.

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/IActionSelector.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/IActionSelector.cs]: YEAH

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/IActionSelector.cs [hunk 3]
```diff
@@ -9,6 +8,6 @@ namespace Microsoft.AspNet.Mvc.Infrastructure
 {
     public interface IActionSelector
     {
-        Task<ActionDescriptor> SelectAsync(RouteContext context);
+        ActionDescriptor Select(RouteContext context);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/IActionSelector.cs]: docs while you're at it?

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/MvcRouteHandler.cs [hunk 4]
```diff
@@ -121,15 +117,22 @@ private void EnsureServices(HttpContext context)
                 return;
             }
 
+            var services = context.RequestServices;
+
+            // Verify if AddMvc was done before calling UseMvc
+            // We use the MvcMarkerService to make sure if all the services were added.
+            MvcServicesHelper.ThrowIfMvcNotRegistered(services);
+
+
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/MvcRouteHandler.cs]: extra space.

File: test/Microsoft.AspNet.Mvc.Core.Test/Infrastructure/DefaultActionSelectorTests.cs [hunk 5]
```diff
@@ -375,9 +374,9 @@ public class DefaultActionSelectorTests
             context.RouteData.Values.Add("action", "Buy");
 
             // Act
-            var ex = await Assert.ThrowsAsync<AmbiguousActionException>(async () =>
+            var ex = Assert.Throws<AmbiguousActionException>(() =>
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/Infrastructure/DefaultActionSelectorTests.cs]: ` var ex = Assert.Throws<AmbiguousActionException>(() => selector.Select(context));`

**Comment by pranavkm** [comment]
:shipit:

**Pull request closed without merging**